llvm/lib/CodeGen/RegAllocBase.cpp

```diff
@@ -161,6 +163,40 @@ void RegAllocBase::postOptimization() {
   DeadRemats.clear();
 }
 
+void RegAllocBase::cleanupFailedVRegs() {
+  SmallSet<Register, 8> JunkRegs;
+
+  for (Register FailedReg : FailedVRegs) {
+    JunkRegs.insert(FailedReg);
+
+    MCRegister PhysReg = VRM->getPhys(FailedReg);
+    LiveInterval &FailedInterval = LIS->getInterval(FailedReg);
+
+    // The liveness information for the failed register and anything interfering
+    // with the physical register we arbitrarily chose is junk and needs to be
+    // deleted.
+    for (MCRegUnitIterator Units(PhysReg, TRI); Units.isValid(); ++Units) {
+      LiveIntervalUnion::Query &Q = Matrix->query(FailedInterval, *Units);
+      for (const LiveInterval *InterferingReg : Q.interferingVRegs())
+        JunkRegs.insert(InterferingReg->reg());
+    }
+  }
+
+  // TODO: Probably need to set undef on any physreg uses not associated with
+  // a virtual register.
```

Comment:
Aren't there places where we lower things/expand pseudos using physical regs before RegAlloc? We do some weird things for stuff like TLSDESC  where ABI dictates we use particular registers. It's not a problem in this patch, but I want to understand better if we're doing things we shouldn't in the various lowerings and should consider addressing that.

Reply:
It depends. There are some that use reserved physical registers, and some that don't. The verifier rules are different. Reserved registers get very little checking but you're not allowed to set most of these operand flags when using them